accepted/00097-recurring-actions.md

```diff
@@ -0,0 +1,321 @@
+2) Delete the schedule if the lookup fails in the `RecurringActionApplyJob` taskomatic job.
+3) Create a trigger to handle it.
+
+This is how a trigger for the example of a linked CLM project could look like:
+```
+CREATE OR REPLACE FUNCTION suse_contentproject_del_trig() RETURNS TRIGGER AS $$
+DECLARE project_id int;
+BEGIN
+  project_id = OLD.id;
+  DELETE FROM suseRecurringActionType
+    WHERE action_type in ('clm-build','clm-promote')
+      AND (SELECT (type_data->>'project_id')::int FROM suseRecurringActionType) = project_id;
+
+  RETURN NULL;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE TRIGGER suse_contentproject_del_trig
+AFTER DELETE ON suseContentProject
+FOR EACH ROW EXECUTE PROCEDURE suse_contentproject_del_trig();
+```
+
+The preferred solution in this case is option 1).
+
+See ERM model below:
+
+![Recurring Action Type table](images/00097-recurring-actions-erm.png)
+
+See [here](https://github.com/uyuni-project/uyuni/pull/6678/commits/ba13bb918c0220a9455ddf1cf2d4373c12742d3b) for a POC of the proposed entity changes.
+
+**Solution #2:**
+
+Another possible solution is to use the `rhnAction` table to store the recurring action data.
+We would create an `rhnAction` entry per recurring action without an associated `serverAction` this way the action won't get picked up by any entity
+and stick around indefinitely. Since recurring actions schedule actions for a specific action type e.g. `ApplyStatesAction` for highstate / custom state we can store all the needed data in the `rhnAction<ACTION_TYPE> ` tables.
+
+See [here](https://github.com/parlt91/uyuni/commit/8ffe1c9e79899c0eb0b58da196ca62999385b06d) for a POC that makes the current highstate implementation use the `rhnAction` to store it's `testMode` parameter.
+
+There is one problem however:
+The recurring `CLM build/promote` action we want to implement don't target a specific entity and thus we don't have an action for them.
+The solution is to create a 'dummy' `ContentManagementAction` that stores the project and the environment we want to build or promote.
+This action only exists for the purpose of storing the data we need to schedule recurrent runs and will never actually be scheduled on an entity (minion/group/org).
+
+See [here](https://github.com/parlt91/uyuni/commit/ac608e7865b0d6e6eee0ff495cc886664489c86d) for a POC of the `ContentManagementAction`.
+
+And [here](https://github.com/parlt91/uyuni/commit/898103d38e194fc1dcf4b72877d566e49f3c799c) for a POC of a working recurring CLM project build backend implementation.
+
+### Backend
+
+See [here](https://github.com/uyuni-project/uyuni/pull/6678/commits/61cadb982fcbdd46546ff219f658aaefa21c1f4b) for an outline on how the use of a specific entity created above would look like for the existing recurring highstate.
+Recurring action type entities would be created/managed based on the action type received from the frontend.
+
+The workflow to add new recurring action types would look like:
```

Comment:
IIUC this workflow is based on Solution 1. Maybe we can rewrite it for solution 2.

Reply:
Done